### PR TITLE
Fix request-map not being cleared up after requests

### LINKED ISSUE #62: Memory leak in setModReq
Version: `master/413781778738c08fdbb98e1dd65f5abffe8832d0`

I noticed that my Go process was consuming more and more memory over time, even though I had an LRU cache set up capped on # bytes. I traced it down using pprof to this library. It appeared that it was holding onto `http.Request` objects created from [cloneRequest](https://github.com/gregjones/httpcache/blob/413781778738c08fdbb98e1dd65f5abffe8832d0/httpcache.go#L539). These requests are being stored in a map, and superficially it looks like the library was clearing them out correctly when the request was over (response read or error occurred), so I spun up gdb:

```
$ gdb -p 11322
(gdb) break httpcache.go:165
Breakpoint 1 at 0x76040c: file /..../vendor/src/github.com/gregjones/httpcache/httpcache.go, line 165.
(gdb) continue

Continuing.
[Switching to Thread 0x7f6cd1ffb700 (LWP 11337)]

Breakpoint 1, github.com/gregjones/httpcache.(*Transport).setModReq (t=0xc420164910, orig=0xc42001b770, mod=0xc42001b860)
165		if mod == nil {

(gdb) print t.modReq
$5 = map[*net/http.Request]*net/http.Request = {
  [0xc4209de1e0] = 0xc42001af00,
  [0xc4203240f0] = 0xc4203241e0,
  [0xc4209de780] = 0xc4209de870,
  [0xc4209de5a0] = 0xc4209de690,
  [0xc42001b770] = 0xc42001b860, 
  [0xc4203250e0] = 0xc4203251d0,
  [0xc42001bc20] = 0xc42001bd10,
  [0xc420324f00] = 0xc420324ff0,
  [0xc4203242d0] = 0xc4203243c0,
  [0xc42001a690] = 0xc42001a780,
  [0xc420325950] = 0xc420325a40}

(gdb) step
166			delete(t.modReq, orig)

(gdb) print orig
$6 = (struct net/http.Request *) 0xc420325a40
```

Notice the problem? `orig` is not a valid **key** in the `map`, it's a **value**. This happens because of [this line](https://github.com/gregjones/httpcache/blob/413781778738c08fdbb98e1dd65f5abffe8832d0/httpcache.go#L228) which is then being passed to `setModReq` [here](https://github.com/gregjones/httpcache/blob/413781778738c08fdbb98e1dd65f5abffe8832d0/httpcache.go#L232) and [here](https://github.com/gregjones/httpcache/blob/413781778738c08fdbb98e1dd65f5abffe8832d0/httpcache.go#L238) which are now referencing `req2`, NOT `req`.

This means that `t.setModReq(req, nil)` is completely broken and will never delete any `http.Requests`, leaking them forever.

### PR DESCRIPTION
**Pull request by gregjones**
Fixes #62 

I think this will do until sorting out the more up-to-date cancelling can happen.

cc @Kegsay @shurcooL

### CODE REVIEW

File: httpcache.go [hunk 1]
```diff
@@ -225,17 +225,18 @@ func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error
 					// Associate original request with cloned request so we can refer to
 					// it in CancelRequest()
 					t.setModReq(req, req2)
+					keyReq := req
```
Comment (shurcooL) [inline httpcache.go]: What does the name of variable `keyReq` mean? Is it the "request that is the _key_ of the `modReq` map"?

I think this code is quite subtle and not obvious, so an inline comment explaining the purpose of this line would be very helpful. Something like this:

```Go
keyReq := req // Keep a pointer to the original request so we can delete it from the `modReq` map later.
```

File: httpcache_test.go [hunk 2]
```diff
@@ -465,6 +465,11 @@ func TestGetWithLastModified(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	defer func() {
+		if len(s.transport.modReq) != 0 {
+			t.Errorf("Request-map is not empty")
+		}
+	}()
```
Comment (shurcooL) [inline httpcache_test.go]: It's great to have a test for this bug. 👍

**Comment by gregjones** [comment]
I like your suggestion, thanks 👍

**Comment by Kegsay** [comment]
Any ETA on this landing? My project's memory usage has doubled in size since Monday (~500MB -> ~1GB) so I really want to apply this fix as soon as possible 🙏

**Pull request merged**